### PR TITLE
Add the primerdb

### PR DESCRIPTION
**Pull request by aledj2**
This change is [Reviewable](https://reviewable.io/reviews/woook/moka-fe/324)

### CODE REVIEW

File: code/PrimerDB/ImportPrimer.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+import os
+
+# import from config file
+from ImportPrimerconfig import *
+
+class ImportPrimer():
+	def __init__(self):
+		# using pyodbc specify database connection details
+		#self.cnxn = pyodbc.connect(mokadata_connection) # use the mokadata database connection details specified in the config file
+		self.cnxn = pyodbc.connect(dev_database_connection) # use the devdatabase database connection details specified in the config file
```
Comment (andyb3) [inline code/PrimerDB/ImportPrimer.py]: The script is still pointing at the dev database. See #347

File: code/PrimerDB/ImportPrimer.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+								# each gene in the bedfile dict has a tuple key of (chr,start,stop), the entrez gene id is the value
+								# for each gene in Pan493 look if the primer overlaps with the gene 
+								if gene[0]==chr.replace("chr","") and gene[1]<primer_stop and gene[2]>primer_start:
+									# capture the entrez gene id
+									entrez_gene_id_mapped=self.bedfile_dict[gene]
+							# NB if a primer overlaps with multiple genes only one will be captured
+							
+							
+							#create a dict of all variables to be inserted to the database and pass to self.build_insert module
+							primer_values_to_insert_dict={"chr":str(cleaned_chromosome_number),"start":str(primer_start),"stop":str(primer_stop),"gene":str(entrez_gene_id_mapped),"FSeq":forward_sequence,"RSeq":reverse_sequence,"FTag":f_tag,"RTag":r_tag,"status":self.status_id,"purification":self.purification_id,"SoS":self.scale_of_synth_id,"GenomeBuild":self.genome_build_id,"notes":variant_id}
```
Comment (andyb3) [inline code/PrimerDB/ImportPrimer.py]: Lines 168 and 200 are very long, consider splitting over multiple lines? See #348

**Comment by andyb3** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[code/PrimerDB/ImportPrimer.py, line 192 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyL9y-1-mgTm3H2OB_w:-KyL9y-1-mgTm3H2OB_x:b-4czote) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/code/PrimerDB/ImportPrimer.py#L192)):*
> ```Python
> 		A dictionary of values to be inserted is recieved and a SQL insert statement built and executed
> 		"""
> 		# The primer name is an autoincrementing number. A select query (ordered by primer name descending) returns the current highest number. this is a union query returning 0 should this be the first import
> ```

Could the purpose of the union query returning 0 be explained a little more clearly? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[code/PrimerDB/ImportPrimer.py, line 19 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyL_Q5sjemLy2yw6WU4:-KyL_Q5sjemLy2yw6WU5:boqm6tp) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/code/PrimerDB/ImportPrimer.py#L19)):*
> ```Python
> 
> # import from config file
> from ImportPrimerconfig import *
> ```

Might it be better to change this import to something like 'import ImportPrimerconfig as c' and then prefix variables from the file with c.<variable>? That way it's clearer in the code that variables have come from the config file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Reviewed 1 of 1 files at r2.
Review status: 1 of 8 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by aledj2** [comment]
Review status: 0 of 8 files reviewed at latest revision, 7 unresolved discussions.

---

*[code/PrimerDB/ImportPrimer.py, line 19 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyL_Q5sjemLy2yw6WU4:-KyS0qZtKhWkau3-ycUf:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/code/PrimerDB/ImportPrimer.py#L19)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Might it be better to change this import to something like 'import ImportPrimerconfig as c' and then prefix variables from the file with c.<variable>? That way it's clearer in the code that variables have come from the config file.
</blockquote></details>

Done.

---

*[code/PrimerDB/ImportPrimer.py, line 168 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyHFlovikN7kglYF_MV-r2-168:-KyV0dLW5BL3lBmgVqqU:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/code/PrimerDB/ImportPrimer.py#L168)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Lines 168 and 200 are very long, consider splitting over multiple lines? See #348
</blockquote></details>

Done.

---

*[code/PrimerDB/ImportPrimer.py, line 192 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyL9y-1-mgTm3H2OB_w:-KyV0eMwDBSfAHZZfHe1:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/code/PrimerDB/ImportPrimer.py#L192)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Could the purpose of the union query returning 0 be explained a little more clearly? 
</blockquote></details>

Done.

---

*[forms/61001_PrimerDB_dashboard, line 8 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyQYbKxZvo7gjT9dbfR:-KyQYbKxZvo7gjT9dbfS:b-aflr8e) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/forms/61001_PrimerDB_dashboard#L8)):*
> ```
>     ' settings to have cmd box visible while running
>     Dim waitOnReturn As Boolean: waitOnReturn = True
>     Dim windowStyle As Integer: windowStyle = 1
> ```

what is this?

---

*[forms/61001_PrimerDB_dashboard, line 12 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyQYZt-lbjrCY_WIcXg:-KyQYZt-lbjrCY_WIcXh:bkxp7w5) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/forms/61001_PrimerDB_dashboard#L12)):*
> ```
>     'allow capture of error code
>     Dim ErrorCode As Integer
>     
> ```

add comment here

---

*[forms/61001_PrimerDB_dashboard, line 20 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyQYX7ymthJ4yqUOmsS:-KyQYX7ymthJ4yqUOmsT:bn4rhj8) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/forms/61001_PrimerDB_dashboard#L20)):*
> ```
>     
>     'combine all to make totalCmd NB /K can be used for the cmd line window to remain after it has run - good for debugging - but may cause it to hang
>     ' note 2>&1 which combines stderr and stdout
> ```

add to comments to explain this command better

---

*[forms/61001_PrimerDB_dashboard, line 42 at r2](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyQYq8DkWGeAi2CEtwm:-KyQYq8DkWGeAi2CEtwn:b-r237b0) ([raw file](https://github.com/woook/moka-fe/blob/8180f6b917e568ff57893364873986d188b967fa/forms/61001_PrimerDB_dashboard#L42)):*
> ```
> End Sub
> 
> Private Sub Command15_Click()
> ```

rename button and add comment to say not currently is use (comment all out?)
Add comments throughout

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Reviewed 1 of 8 files at r1, 4 of 6 files at r3.
Review status: 5 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[forms/61006_FindPrimerByID, line 9 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW4zH4hF9uBg_4CzEe:-KyW4zH4hF9uBg_4CzEf:b497rv7) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/61006_FindPrimerByID#L9)):*
> ```
> QRY = "SELECT DISTINCT PrimerAmplicon.PrimerName, Chromosome.Chr, PrimerAmplicon.Start, PrimerAmplicon.Stop, PrimerTags.Name AS ForwardTag, dbo_PrimerTags_1.Name AS ReverseTag, PrimerAmplicon.Notes, PrimerAmplicon.ForwardSeq, PrimerAmplicon.ReverseSeq, PrimerLookup.Name AS GenomeBuild, PrimerAmplicon.Mix, PrimerAmplicon.RTray, PrimerAmplicon.RFreezer, PrimerAmplicon.FGrid, PrimerAmplicon.FTray, PrimerAmplicon.FFreezer, PrimerAmplicon.TestResultNotes, PrimerAmplicon.Status" & _
> " FROM (((PrimerAmplicon INNER JOIN Chromosome ON PrimerAmplicon.ChromosomeID = Chromosome.ChrID) INNER JOIN PrimerTags ON PrimerAmplicon.FTag = PrimerTags.PrimerTagsID) INNER JOIN PrimerTags AS dbo_PrimerTags_1 ON PrimerAmplicon.RTag = dbo_PrimerTags_1.PrimerTagsID) INNER JOIN PrimerLookup ON PrimerAmplicon.GenomeBuild = PrimerLookup.PrimerLookupID" & _
> " WHERE PrimerAmplicon.PrimerName = " & [Forms]![61006_FindPrimerByID]![FindAmpliconID] & " AND PrimerAmplicon.Status=1202218832;"
> ```

What is status 1202218832 found in the  WHERE clause of query? Could you state it in the comments?

---

*[forms/61007_EnterNewPrimer, line 1 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW3LgmfgMGJLyMfGpG:-KyW3LgmfgMGJLyMfGpH:bc46o4l) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/61007_EnterNewPrimer#L1)):*
> ```
> Option Compare Database
> ```

Could some comments be added to this form?

---

*[forms/s61003_ViewAllAwaitingOrdering, line 5 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW2iS9_lChhcM8aJLk:-KyW2iSAH-TtTh1euQot:b-tacgcl) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/s61003_ViewAllAwaitingOrdering#L5)):*
> ```
> Private Sub ID_DblClick(Cancel As Integer)
> ' this function is run when doubleclicking the ID column
> ' the function takes all primers which are awaiting ordering and creates a text file which can be used to upload to the primer ordering website
> ```

This says it creates a text file but isn't it an Excel spreadsheet?

---

*[queries/s61003_PrimerOrderSequence, line 1 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW3w7M0ponTjBDG6HN:-KyW3w7M0ponTjBDG6HO:b-8cbisn) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/queries/s61003_PrimerOrderSequence#L1)):*
> ```
> SELECT PrimerAmplicon.PrimerName & "F_" & PrimerTags.Ordering_name & "_01" AS PrimerName, "" AS [5' Mod], [Sequence] & [ForwardSeq] AS [Oligo Sequence], "" AS [3' Mod], PrimerLookup.Name AS ScaleOfSynth, PrimerLookup_1.Name AS Purification, "In solution (TE)" AS [In Solution], 200 AS 200, 1 AS 1, "Dilute to 250" AS Notes
> ```

Could you change the '200' and '1' fields to a descriptive name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Reviewed 1 of 6 files at r3.
Review status: 6 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[forms/61006_FindPrimerByID, line 9 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW4zH4hF9uBg_4CzEe:-KyW6SbjIN72zYOKLp2J:bcvh6od) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/61006_FindPrimerByID#L9)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

What is status 1202218832 found in the  WHERE clause of query? Could you state it in the comments?
</blockquote></details>

Just saw in form 61002 the status is 'tested in use'. Could you add that into the comments on here too? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Reviewed 1 of 6 files at r3.
Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by aledj2** [comment]
Review status: 4 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[forms/61006_FindPrimerByID, line 9 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW4zH4hF9uBg_4CzEe:-KyW8vB84nV8M7j09sm8:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/61006_FindPrimerByID#L9)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Just saw in form 61002 the status is 'tested in use'. Could you add that into the comments on here too? 
</blockquote></details>

Done.

---

*[forms/61007_EnterNewPrimer, line 1 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW3LgmfgMGJLyMfGpG:-KyWAHtlhjErqBqfvMEF:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/61007_EnterNewPrimer#L1)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Could some comments be added to this form?
</blockquote></details>

Done.

---

*[forms/s61003_ViewAllAwaitingOrdering, line 5 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW2iS9_lChhcM8aJLk:-KyWAhnI4jUqIEyHgq5Z:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/forms/s61003_ViewAllAwaitingOrdering#L5)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

This says it creates a text file but isn't it an Excel spreadsheet?
</blockquote></details>

Done.

---

*[queries/s61003_PrimerOrderSequence, line 1 at r3](https://reviewable.io:443/reviews/woook/moka-fe/324#-KyW3w7M0ponTjBDG6HN:-KyWBr8DXg-oR7dkQGYn:b-896fix) ([raw file](https://github.com/woook/moka-fe/blob/13ded78f4e611260d314f8d911daca79c207aa4d/queries/s61003_PrimerOrderSequence#L1)):*
<details><summary><i>Previously, andyb3 wrote…</i></summary><blockquote>

Could you change the '200' and '1' fields to a descriptive name?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Comment by andyb3** [comment]
Reviewed 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/woook/moka-fe/324)*

**Pull request merged**